extra_data/reader.py

```diff
@@ -581,17 +586,340 @@ def union(self, *others):
+
+            # Use the standard library's tomllib if possible and try
+            # tomli if necessary.
+            if py_version.major >= 3 and py_version.minor >= 11:
+                from tomllib import load as load_toml
+            else:
+                from tomli import load as load_toml
+
+            with open(aliases_path, 'rb') as f:
+                data = load_toml(f)
+
+        aliases = {}
+
+        def walk_dict_value(source, key_aliases):
+            for alias, key in key_aliases.items():
+                aliases[alias] = (source, key)
+
+        for key, value in data.items():
+            if isinstance(value, str):
+                # Source alias.
+                aliases[key] = value
+            elif isinstance(value, list) and len(value) == 2:
+                # Sourcekey alias by explicit list.
+                aliases[key] = tuple((str(x) for x in value))
+            elif isinstance(value, dict):
+                # Sourcekey alias by nested mapping.
+                walk_dict_value(key, value)
+            else:
+                raise ValueError(f"unsupported literal type for alias '{key}'")
+
+        return aliases
+
+    def with_aliases(self, *alias_defs):
+        """Apply aliases for convenient source and key access.
+
+        Allows to define aliases for sources or source-key combinations
+        that may be used instead of their literal names to retrieve
+        :class:`SourceData` and :class:`KeyData` objects via
+        :property:`DataCollection.alias`.
+
+        Multiple alias definitions may be passed as positional arguments
+        in different formats:
+
+        1. Passing a dictionary mapping aliases to sources (passed as strings)
+           or source-key pairs (passed as a 2-len tuple of strings).
+
+        2. Passing a string or PathLike pointing to a JSON, YAML (requires
+           pyYAML installed) or TOML (requires Python 3.11 or with tomli
+           installed) file containing the aliases. For unsupported formats,
+           an ImportError is raised.
+
+           The file should contain mappings from alias to sources as strings
+           or source-key pairs as lists. In addition, source-key aliases may
+           be defined by nested key-value pairs according to the respective
+           format, shown here in YAML::
+
+             # Source alias.
+             sa1-xgm: SA1_XTD2_XGM/XGM/DOOCS
+
+             # Direct source key alias.
+             sa1-intensity: [SA1_XTD2_XGM/XGM/DOOCS:output, data.intensityTD]
+
+             # Nested source key alias.
+             SA3_XTD10_MONO/MDL/PHOTON_ENERGY:
+               mono-central-energy: actualEnergy
+
+        Returns a new :class:`DataCollection` object with the aliases
+        for sources and keys.
+        """
+
+        return DataCollection(
+            self.files, sources_data=self._sources_data,
+            train_ids=self.train_ids, aliases=self._parse_aliases(alias_defs),
```

Comment:
Should using `.with_aliases()` a second time discard your existing aliases or add to them? I was imagining it would add to them, but I'm not set on that.

E.g. I could imagine that you might have some shared aliases from the instrument or the proposal, plus some local ones specific to a particular analysis. You could apply both in one call to `.with_aliases()`, of course, but maybe it would feel more natural to do them in separate steps. :shrug:

Reply:
Amusingly I didn't even intend either behaviour when writing this, even though it may seem like that with `.with_aliases()` accepting multiple definitions at the same time. The only reason for removing aliases I can think of is to avoid conflicts during a merge.

So the options are keeping `.with_aliases()` accepting multiple inputs and replacing, or having it extend aliases probably taking ony a single definition, and add some `.strip_aliases()` method to remove them.

I don't have a particular good argument for either, except maybe the other `.select()`/`.deselect()` methods by nature cannot be chained in a "merging" sense, so one may keep this idiom here.

Reply:
If we do add automatic loading of aliases from a proposal number, I think that pushes me a bit towards saying that `.with_aliases()` could be additive, so that `open_run(3124, 11).with_aliases(...)` can combine the shared proposal aliases with your additional ones. I'm still not 100% set on that if there's a good argument to the contrary, but leaning that way.

Reply:
I see the point, it'd be confusing to use the automatic aliases and suddenly lose them when adding more custom aliases.

The functionality of an additive `.with_aliases()` is present in this version though, albeit clunky:

```
dc = dc.merge(dc.with_aliases('additional-aliases.toml'))
```
Moving this functionality to `.with_aliases` begs the question how it should handle collisions? At the moment those in between the individual alias definitions are ignored, while `.merge()` checks for conflicts.

If we handle conflicts on some though, someone likely would like to remove aliases to add entirely different ones, so we need a give-me-this-DataCollection-without-aliases function as well.

Reply:
> Moving this functionality to `.with_aliases` begs the question how it should handle collisions?

Good question. My programmer instinct is to say 'last source wins', like `dict.update()`, but it would clearly get confusing if some code expects 'motor1' to be one source and you're redefining it to another source. So maybe checking for conflicts is the right way to go. Redefining the alias pointing to the same name should be OK, though (as I think it is in the current `.merge()` implementation).

 > If we handle conflicts on some though, someone likely would like to remove aliases to add entirely different ones, so we need a give-me-this-DataCollection-without-aliases function as well.

Makes sense. `.drop_aliases()`? `.without_aliases()`?

Reply:
> My programmer instinct is to say 'last source wins'

I wonder, how does this case differ from the discussion in https://github.com/European-XFEL/EXtra-data/pull/298 regarding order of overwriting? As a matter of fact, `.with_aliases()` worked like that up to now and simply overwrote whatever aliases are defined, in order.

I wonder, how about we make `.with_aliases()` simply run the line merge line above when it's called on a `DataCollection` that already has aliases defined?

Reply:
I suspect there are more efficient ways to do it than `.union()`, because that does a bunch of set operations on files and train IDs that are unnecessary for this. But it's probably not going to be a massive difference, so if that's the easiest way to get it working, it's probably fine.

> I wonder, how does this case differ from the discussion in https://github.com/European-XFEL/EXtra-data/pull/298 regarding order of overwriting?

I wondered if you'd think of that. :wink: I don't have a satisfying answer, but my mental model seemingly goes differently if I'm envisaging a single pool of data (last update wins) or some kind of layers (first=highest wins). In this case, I think that uncertainty is a good argument for error-ing on conflicts and letting the user clear it up.

Reply:
I gave this another thought and before implementation would like to re-check our plan:

* `DataCollection.with_aliases` merges any new alias definitions into the existing ones (erroring out on conflicts) on the `DataCollection`
* `DataCollection.only_aliases` merges any new alias definitions into the existing ones (erroring out on conflicts) and selects down to **all** aliases then defined on the `DataCollection`
* `DataCollection.merge` works as before and identical to `DataCollection.with_aliases` with respect to aliases
* Add a new call `DataCollection.drop_aliases`

Reply:
Yes, given the alias API is somewhat intended to "hide" the true source, I would avoid using any form of overwriting scheme and visibly error out if there's a conflict.

I changed the behaviour of all calls now to always add aliases instead of replacing them. When implementing the conflict detection, I noticed the previous implementation of checking for those was actually not complete - if you merged multiple `DataCollections` using `.merge()` and the conflict was only in between the added ones rather than the one you call it from, it would not have been detected. The new code now active for all alias operations properly checks for any combination.